### DISCUSSION TITLE
Divergence of divergence of a tensor

### DISCUSSION THREAD

**Discussion by alexguev** [idaholab/moose#20948] [Q&A General] [answered]
Hi,

I would like to implement in a kernel the term ![](https://latex.codecogs.com/svg.latex?\Large&space;\nabla\cdot(\lambda\nabla\cdot\sigma)), where ![](https://latex.codecogs.com/svg.latex?\Large&space;\sigma=\kappa\nabla\varphi\otimes\nabla\varphi) (a capillary stress), where I solve for the variable ![](https://latex.codecogs.com/svg.latex?\Large&space;\varphi). Note that ![](https://latex.codecogs.com/svg.latex?\Large&space;\lambda) depends on ![](https://latex.codecogs.com/svg.latex?\Large&space;\varphi).

I think that I have managed to implement the computation of the stress through a Material following ComputeSurfaceTensionKKS, here is the relevant code:

```
void
ADCapillaryStress::computeQpProperties()
{
  auto & S = _capillary_stress[_qp];
  S.zero();

  const ADReal nx = _grad_v[_qp](0);
  const ADReal ny = _grad_v[_qp](1);
  const ADReal nz = _grad_v[_qp](2);

  S(0, 0) += _kappa[_qp] * nx * nx;
  S(0, 1) += _kappa[_qp] * nx * ny;
  S(1, 1) += _kappa[_qp] * ny * ny;
  S(0, 2) += _kappa[_qp] * nx * nz;
  S(1, 2) += _kappa[_qp] * ny * nz;
  S(2, 2) += _kappa[_qp] * nz * nz;

  // fill in symmetrically
  S(1, 0) = S(0, 1);
  S(2, 0) = S(0, 2);
  S(2, 1) = S(1, 2);
}
```

Now if the divergence of the stress were a gradient, this would be the usual diffusion equation, using _grad_u:

```
ADReal
ADAnisotropicDiffusion::computeQpResidual()
{
  return _lambda[_qp] * _grad_u[_qp] * _grad_test[_i][_qp];
}
```

But here instead of _grad_u, in my weak form, I have divergence of the stress. I have noticed grad.row of potential relevance in InitialConditionTempl.C, but not sure at all. Taking inspiration from ADStressDivergenceTensors, maybe something like that (probably wrong though):

```
ADReal
ADAnisotropicDiffusion::computeQpResidual()
{
  return _lambda[_qp] * _stress[_qp].grad.row(_component) * _grad_test[_i][_qp];
}
```

The other possibility would be to deport the two derivatives onto the test function, but I would rather avoid using higher order elements.

Thanks

**Comment by hugary1995** [comment, marked as answer]
You need third order spatial derivatives according to the equations you showed. So you can't really avoid using C1 function space. Hermite polynomials is one choice.

**Reply by hugary1995** [reply]
Correction: you could still use C0 functions if you use a split formulation, i.e. define some dummy variables to reduce the regularity of the stress divergence term, similar to the split cahn Hilliard equations.

**Reply by alexguev** [reply]
Hey Gary, thanks for the prompt reply. 

Yep I was thinking of doing the split as for Cahn-Hilliard equations (strategy that I use already on a chemical potential as I'm doing Cahn-Hilliard for another part of the equation). My main concern is how to replace _grad_u[_qp] by the divergence of a tensor.

**Reply by hugary1995** [reply]
I wrote down some equations to explain:
![image](https://user-images.githubusercontent.com/31555204/166845696-864b6bf6-d0d3-4896-a8ad-c8889d4df9ef.png)

**Reply by alexguev** [reply]
Noted, thanks! I'll try take it from there.
